src/transformers/models/gemma2/modeling_gemma2.py

```diff
@@ -807,7 +807,19 @@ def forward(
             inputs_embeds = self.embed_tokens(input_ids)
 
         if cache_position is None:
-            cache_position = torch.arange(0, inputs_embeds.shape[1], device=inputs_embeds.device)
+            if past_key_values is None:
+                cache_position = torch.arange(0, inputs_embeds.shape[1], device=inputs_embeds.device)
+            else:
+                raise ValueError("When `past_key_values` is passed, `cache_position` must be too")
+
+        # Probably a forward call with caching, so we set up cache for one call only
+        if use_cache and past_key_values is None and not self.training:
+            logger.warning_once(
+                "You are calling the model with `use_cache=True` but didn't pass `past_key_values`. ",
+                "Make sure to pass an instance of `HybridCache`. Caching will be disabled. See for more: "
+                "(https://huggingface.co/docs/transformers/main/en/internal/generation_utils#transformers.HybridCache)",
+            )
+            use_cache = False
```

Comment:
TBH I am bothered by adding a LOT of if else if else and stuff like that, which should be more often than not handle outside . Tho I guess we don't have a choice

Reply:
what if we change the good old 
```python3 
use_cache = use_cache if use_cache is not None else self.config.use_cache || not self.training

Reply:
We could do, the main concern is mentioned by user in the issue, that people expect gemma2 to be as all other models and return cache if use_cache is True. If we simply change the `use_cache` it won't make it more explicit for those not familiar with cache classes. So my main point is to make users familiar that gemma2 is different and they should not expect same behavior as in other models.

Reply:
Got it. There a tradeoff to have between people's expectations and warning everywhere. IMO we should:
- better document Gemma2: add in the gemma2.md that gemma has a different cache class
- return the HybridCache only when you have `use_cache` / :  
```python
        return_legacy_cache = False
        if (
            use_cache and not isinstance(past_key_values, Cache) and not self.training
        ):  # kept for BC (non `Cache` `past_key_values` inputs)
            return_legacy_cache = True
            past_key_values = DynamicCache.from_legacy_cache(past_key_values)
            logger.warning_once(
                "We detected that you are passing `past_key_values` as a tuple and this is deprecated and will be removed in v4.43. "
                "Please use an appropriate `Cache` class (https://huggingface.co/docs/transformers/v4.41.3/en/internal/generation_utils#transformers.Cache)"
            )


```
In llama we already have this, meaning if you past_key_value is None we only create dynamic cache when not training

Reply:
Yes, we should also document it, will add some info. So, init an empty HybridCache should be okay, as long as we warn and document it